core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ImplementTableFunctionSource.java

```diff
@@ -0,0 +1,500 @@
+        // In the new map, every source symbol is associated with the corresponding marker symbol.
+        // Null value of the marker indicates that the source value should be ignored by the table function.
+        ImmutableMap<Symbol, Symbol> markerSymbols = rowNumberSymbols.entrySet().stream()
+                .collect(toImmutableMap(Map.Entry::getKey, entry -> marked.getSymbolToMarker().get(entry.getValue())));
+
+        // Keep the final row number symbol, and use it for ordering the combined sources.
+        // It runs along each partition in the cartesian product, numbering the partition's rows according to the expected ordering / orderings.
+        // note: ordering is necessary even if all the source tables are not ordered. Thanks to the ordering, the original rows
+        // of each input table come before the "filler" rows.
+        Optional<OrderingScheme> finalOrderBy = Optional.of(new OrderingScheme(ImmutableList.of(finalRowNumberSymbol), ImmutableMap.of(finalRowNumberSymbol, ASC_NULLS_LAST)));
+
+        // Partition the combined sources by their combined partitioning lists.
+        List<Symbol> partitionBy = node.getTableArgumentProperties().stream()
+                .map(TableArgumentProperties::getSpecification)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .map(DataOrganizationSpecification::getPartitionBy)
+                .flatMap(Collection::stream)
+                .collect(toImmutableList());
+
+        PlanNode pruned = pruneHelperSymbols(marked.getNode(), partitionSizeSymbols.build(), rowNumberSymbols.values(), combinedRowNumberSymbols.build(), finalRowNumberSymbol, context);
+
+        // Combine the pass through specifications from all sources
+        List<PassThroughSpecification> passThroughSpecifications = node.getTableArgumentProperties().stream()
+                .map(TableArgumentProperties::getPassThroughSpecification)
+                .collect(toImmutableList());
+
+        // Combine the required symbols from all sources
+        List<List<Symbol>> requiredSymbols = node.getTableArgumentProperties().stream()
+                .map(TableArgumentProperties::getRequiredColumns)
+                .collect(toImmutableList());
+
+        return Result.ofPlanNode(new TableFunctionProcessingNode(
+                node.getId(),
+                node.getName(),
+                node.getProperOutputs(),
+                Optional.of(pruned),
+                passThroughSpecifications,
+                requiredSymbols,
+                Optional.of(markerSymbols),
+                Optional.of(new DataOrganizationSpecification(partitionBy, finalOrderBy)),
+                ImmutableSet.of(),
+                0,
+                Optional.empty(),
+                node.getHandle()));
+    }
+
+    private static NodeWithSymbols appendHelperSymbolsForSource(
+            int index,
+            TableFunctionNode node,
```

Comment:
Instead of passing the node and the source index, pass the table argument properties, the source node. That way, this method doesn't need to be aware of the TableFunctionNode and its structure.